### PR TITLE
deprecate(core): removed the site_guid entity attribute from datamodel

### PR DESCRIPTION
**Pull request by jdalsem**
This is WIP for dropping multisite features #7372
- [x] Write upgrading docs
- [x] Test fresh install
- [x] ~~Decide if we want to drop function parameters or just ignore them (so there is less breaking)~~ just ignore them
- [x] Write upgrade script to drop columns
- [x] ~~do we offer a migration process if you have multiple sites???~~ personally i decided not to do this

BREAKING CHANGE:
This breaks a lot of site_guid related features, like all the
elgg_get_entities functions. Entities will no longer have a site_guid
attribute.

See http://learn.elgg.org/upgrading#Elgg3.0 for more details on all the
deprecated features regarding this change

### CODE REVIEW

**Comment by jdalsem** [comment]
This is just an early WIP... but with these changes my default Elgg site still functions :smile: 

I checked how much impact this would have with our plugins and that would be very little. We hardly use site_guid in functions or in elgg_get_entities functions...

**Comment by mrclay** [comment]
Wow! I'm definitely for not breaking APIs if we can avoid it; but throw an exception if someone passes in a non-1 integer. Of the goals here, I'd put simplifying/speeding up queries at the top, and saving disk space by altering tables at the bottom. As long as it doesn't get in the way, legacy schema and API is OK to keep around; we could just mark it with comments that it will always be `1`. My 2 cents anyway.

**Comment by mrclay** [comment]
I've not yet digested everything here BTW, but I'm excited it's being considered.

File: engine/lib/configuration.php [hunk 1]
```diff
@@ -88,12 +87,11 @@ function elgg_get_engine_path() {
...
  *
  * @param string $name      Name of the configuration value
- * @param int    $site_guid null for installation setting, 0 for default site
- *
+ * @param int    $site_guid null for installation setting anything else for current site
  * @return mixed Configuration value or null if it does not exist
  * @since 1.8.0
  */
-function elgg_get_config($name, $site_guid = 0) {
+function elgg_get_config($name, $site_guid = null) {
```
Comment (hypeJunction) [inline engine/lib/configuration.php]: Just remove the argument.

File: engine/classes/Elgg/Config.php [hunk 2]
```diff
@@ -149,23 +138,21 @@ public function getCachePath() {
 	 * {@inheritdoc}
 	 */
 	public function get($name, $site_guid = 0) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Config.php]: Remove the argument

File: engine/classes/Elgg/Config.php [hunk 3]
```diff
@@ -149,23 +138,21 @@ public function getCachePath() {
...
 	
-		// do not return $config value if asking for non-current site
-		if (($site_guid === 0 || $site_guid === null || $site_guid == $this->config->site_guid) && isset($this->config->$name)) {
+		if (isset($this->config->$name)) {
 			return $this->config->$name;
 		}
 	
 		if ($site_guid === null) {
 			// installation wide setting
 			$value = _elgg_services()->datalist->get($name);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Config.php]: Just return this and be done with it.

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 4]
```diff
@@ -100,14 +100,14 @@ public function __construct(
...
-	 * @param int  $site_guid Site ID; defaults to current site
+	 * @param int  $ignored   Unused parameter
 	 * @param bool $flush     If set to true, will refresh the access list from the
 	 *                        database rather than using this function's cache.
 	 *
 	 * @return string A list of access collections suitable for using in an SQL call
 	 * @access private
 	 */
-	function getAccessList($user_guid = 0, $site_guid = 0, $flush = false) {
+	function getAccessList($user_guid = 0, $ignored = 0, $flush = false) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Private API, just remove the argument and update it's uses everywhere.

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 5]
```diff
@@ -160,13 +157,13 @@ function getAccessList($user_guid = 0, $site_guid = 0, $flush = false) {
...
 	 * @param int  $user_guid User ID; defaults to currently logged in user
-	 * @param int  $site_guid Site ID; defaults to current site
+	 * @param int  $ignored   Ignored parameter
 	 * @param bool $flush     If set to true, will refresh the access ids from the
 	 *                        database rather than using this function's cache.
 	 *
 	 * @return array An array of access collections ids
 	 */
-	function getAccessArray($user_guid = 0, $site_guid = 0, $flush = false) {
+	function getAccessArray($user_guid = 0, $ignored = 0, $flush = false) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Remove argument and update uses.

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 6]
```diff
@@ -444,13 +435,13 @@ public function hasAccessToEntity($entity, $user = null) {
...
 	 * @param int   $user_guid    The user's GUID.
-	 * @param int   $site_guid    The current site.
+	 * @param int   $ignored      Ignored parameter
 	 * @param bool  $flush        If this is set to true, this will ignore a cached access array
 	 * @param array $input_params Some parameters passed into an input/access view
 	 *
 	 * @return array List of access permissions
 	 */
-	function getWriteAccessArray($user_guid = 0, $site_guid = 0, $flush = false, array $input_params = array()) {
+	function getWriteAccessArray($user_guid = 0, $ignored = 0, $flush = false, array $input_params = array()) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/AccessCollections.php]: Remove argument and update uses

File: engine/classes/Elgg/Database/EntityTable.php [hunk 7]
```diff
@@ -501,7 +498,10 @@ public function enable($guid, $recursive = true) {
 	 */
 	public function getEntities(array $options = array()) {
 
-
+		if (isset($options['site_guid']) || isset($options['site_guids'])) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: I would say just throw an exception if not current site guid

Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: Do this after the pluralization has been done.

File: engine/classes/Elgg/Database/EntityTable.php [hunk 8]
```diff
@@ -1319,17 +1317,12 @@ public function getEntityAttributeWhereSql(array $options = array()) {
...
 	 *
 	 * @return array|false Either an array months as YYYYMM, or false on failure
 	 */
-	public function getDates($type = '', $subtype = '', $container_guid = 0, $site_guid = 0, $order_by = 'time_created') {
-
-		$site_guid = (int) $site_guid;
-		if ($site_guid == 0) {
-			$site_guid = $this->config->get('site_guid');
-		}
+	public function getDates($type = '', $subtype = '', $container_guid = 0, $ignored = 0, $order_by = 'time_created') {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Database/EntityTable.php]: Private API, remove the argument, or better yet update to accept an array of options

File: engine/classes/ElggEntity.php [hunk 9]
```diff
@@ -1323,7 +1321,7 @@ public function removeFromSite($site) {
 	/**
 	 * Gets the sites this entity is a member of
 	 *
-	 * Site membership is determined by relationships and not site_guid.
+	 * Site membership is determined by relationship.
```
Comment (hypeJunction) [inline engine/classes/ElggEntity.php]: This doesn't make sense

Comment (hypeJunction) [inline engine/classes/ElggEntity.php]: Remove this API altogether and wipe the relationships in the upgrade script.

Comment (hypeJunction) [inline engine/classes/ElggEntity.php]: Same for ElggEntity::getSites(), or just return `[elgg_get_site_entity()]`

File: engine/classes/ElggSite.php [hunk 10]
```diff
@@ -221,7 +214,7 @@ public function setDisplayName($displayName) {
 	 * Adds an entity to the site.
 	 *
 	 * This adds a 'member_of_site' relationship between between the entity and
-	 * the site. It does not change the site_guid of the entity.
+	 * the site.
```
Comment (hypeJunction) [inline engine/classes/ElggSite.php]: Just wipe these as well. No point in keeping them around, as all entities belong to a single site

File: engine/lib/access.php [hunk 11]
```diff
@@ -63,15 +63,14 @@ function elgg_get_ignore_access() {
...
  * @param bool $flush     If set to true, will refresh the access list from the
  *                        database rather than using this function's cache.
  *
  * @return string A list of access collections suitable for using in an SQL call
  * @access private
  */
-function get_access_list($user_guid = 0, $site_guid = 0, $flush = false) {
-	return _elgg_services()->accessCollections->getAccessList($user_guid, $site_guid, $flush);
+function get_access_list($user_guid = 0, $flush = false) {
+	return _elgg_services()->accessCollections->getAccessList($user_guid, $flush);
```
Comment (hypeJunction) [inline engine/lib/access.php]: You kept the second attribute in the access collections table.  This won't work.

Comment (hypeJunction) [inline engine/lib/access.php]: This function is used in core plugins. We can keep the signature here, but change in access collections

File: engine/lib/access.php [hunk 12]
```diff
@@ -93,14 +92,13 @@ function get_access_list($user_guid = 0, $site_guid = 0, $flush = false) {
...
- * @param int  $site_guid Site ID; defaults to current site
  * @param bool $flush     If set to true, will refresh the access ids from the
  *                        database rather than using this function's cache.
  *
  * @return array An array of access collections ids
  */
-function get_access_array($user_guid = 0, $site_guid = 0, $flush = false) {
-	return _elgg_services()->accessCollections->getAccessArray($user_guid, $site_guid, $flush);
+function get_access_array($user_guid = 0, $flush = false) {
+	return _elgg_services()->accessCollections->getAccessArray($user_guid, $flush);
```
Comment (hypeJunction) [inline engine/lib/access.php]: Keep the signature, change in access collections. We can then deprecated these and add an elgg_ version with the new signature

File: engine/lib/access.php [hunk 13]
```diff
@@ -258,14 +256,13 @@ function has_access_to_entity($entity, $user = null) {
...
- * @param int   $site_guid    The current site.
  * @param bool  $flush        If this is set to true, this will ignore a cached access array
  * @param array $input_params Some parameters passed into an input/access view
  *
  * @return array List of access permissions
  */
-function get_write_access_array($user_guid = 0, $site_guid = 0, $flush = false, array $input_params = array()) {
-	return _elgg_services()->accessCollections->getWriteAccessArray($user_guid, $site_guid, $flush, $input_params);
+function get_write_access_array($user_guid = 0, $flush = false, array $input_params = array()) {
+	return _elgg_services()->accessCollections->getWriteAccessArray($user_guid, $flush, $input_params);
```
Comment (hypeJunction) [inline engine/lib/access.php]: Same

File: engine/lib/configuration.php [hunk 14]
```diff
@@ -122,12 +120,12 @@ function elgg_set_config($name, $value) {
...
  * @param string $name      Configuration name (cannot be greater than 255 characters)
  * @param mixed  $value     Configuration value. Should be string for installation setting
- * @param int    $site_guid null for installation setting, 0 for default site
+ * @param int    $site_guid null for installation setting anything else defaults to current site
  *
  * @return bool
  * @since 1.8.0
  */
-function elgg_save_config($name, $value, $site_guid = 0) {
+function elgg_save_config($name, $value, $site_guid = null) {
```
Comment (hypeJunction) [inline engine/lib/configuration.php]: Remove argument

File: engine/lib/entities.php [hunk 15]
```diff
@@ -483,15 +481,12 @@ function _elgg_get_entity_attribute_where_sql(array $options = array()) {
...
  * @param string $order_by       Order_by SQL order by clause
  *
  * @return array|false Either an array months as YYYYMM, or false on failure
  */
-function get_entity_dates($type = '', $subtype = '', $container_guid = 0, $site_guid = 0,
-		$order_by = 'time_created') {
-	return _elgg_services()->entityTable->getDates(
-		$type, $subtype, $container_guid, $site_guid, $order_by);
+function get_entity_dates($type = '', $subtype = '', $container_guid = 0, $order_by = 'time_created') {
+	return _elgg_services()->entityTable->getDates($type, $subtype, $container_guid, $order_by);
```
Comment (hypeJunction) [inline engine/lib/entities.php]: Keep the signature, change in entity table

File: engine/lib/tags.php [hunk 16]
```diff
@@ -89,7 +87,6 @@ function elgg_get_tags(array $options = array()) {
 
 		'owner_guids' => ELGG_ENTITIES_ANY_VALUE,
 		'container_guids' => ELGG_ENTITIES_ANY_VALUE,
-		'site_guids' => $CONFIG->site_guid,
```
Comment (hypeJunction) [inline engine/lib/tags.php]: May add `_elgg_validate_site_guid()` that will throw an exception if site guid is not current site guid, and use it in ege* functions?

**Review by hypeJunction — COMMENTED** [review]
I think all SQL clauses as well as the database tables should stay as is, otherwise we end up with hen and egg situation with DB inserts. We just set site guid to 1 on DB writes. We can then loose the columns in the 4.x.

**Comment by jdalsem** [comment]
> I think all SQL clauses as well as the database tables should stay as is, otherwise we end up with hen and egg situation with DB inserts.

what cases do you think off?

File: engine/lib/configuration.php (hunk 1 above)
Comment (jdalsem) [inline engine/lib/configuration.php]: did that before...  think it is better to leave as is... internally the get function acts different using NULL vs any int. NULL gets from datalist... ints from ConfigTable.. did not see a way to do this in an easy way... any idea?

**Comment by hypeJunction** [comment]
If you are going to try and write into a table that sets guid column to NOT NULL, that will fail

**Comment by jdalsem** [comment]
> If you are going to try and write into a table that sets guid column to NOT NULL, that will fail

when (chronological) would that become a problem? Are you thinking after upgrade? Before? During?

**Comment by hypeJunction** [comment]
Exactly. I don't see a proper timeline in this PR.  Part of it is now, the other part is somewhen. Hence I suggest keeping sql queries and schema as is until we know where we are.

File: engine/classes/Elgg/Database/AccessCollections.php (hunk 4 above)
Comment (jdalsem) [inline engine/classes/Elgg/Database/AccessCollections.php]: done

File: engine/classes/Elgg/Database/AccessCollections.php (hunk 5 above)
Comment (jdalsem) [inline engine/classes/Elgg/Database/AccessCollections.php]: done

File: engine/classes/Elgg/Database/AccessCollections.php (hunk 6 above)
Comment (jdalsem) [inline engine/classes/Elgg/Database/AccessCollections.php]: done

File: engine/classes/Elgg/Database/EntityTable.php (hunk 8 above)
Comment (jdalsem) [inline engine/classes/Elgg/Database/EntityTable.php]: done

File: engine/classes/ElggEntity.php (hunk 9 above)
Comment (jdalsem) [inline engine/classes/ElggEntity.php]: relationship does not have to be created at the same time as the entity... so it makes sense... could be discussed in another ticket

File: engine/classes/ElggSite.php (hunk 10 above)
Comment (jdalsem) [inline engine/classes/ElggSite.php]: don't know if the relation was explicitly added for multisite purposes as timing of entity creation and relationship are different (or could be different)

File: engine/lib/access.php (hunk 11 above)
Comment (jdalsem) [inline engine/lib/access.php]: done

File: engine/lib/access.php (hunk 12 above)
Comment (jdalsem) [inline engine/lib/access.php]: done

File: engine/lib/access.php (hunk 13 above)
Comment (jdalsem) [inline engine/lib/access.php]: done

File: engine/lib/entities.php (hunk 15 above)
Comment (jdalsem) [inline engine/lib/entities.php]: done

**Comment by jdalsem** [comment]
Created #10473 to discuss dropping member_of_site relationship

**Comment by hypeJunction** [comment]
In general, I am in favor of dropping as much of the useless API as we can. These ``convenience`` methods create a lot of inconvenience for the core team. Especially stuff like relationships that are easily added and queried, we don't need to keep maintaining it in core.

File: engine/lib/tags.php (hunk 16 above)
Comment (jdalsem) [inline engine/lib/tags.php]: done

**Comment by mrclay** [comment]
What part will get easier to remove in 4.0 and why?

**Comment by hypeJunction** [comment]
I don't know. This PR just feels incomplete, and it provides no overview of what the upgrade path is going to be. It doesn't eliminate all reference to $site_guid, so it feels like it hasn't been thought through. What's going to happen with config table to that has $site_guid in some cases and not others? I would like to have a better full picture, or break this down into multiple steps (4.0 or another PR before 3.0).

**Comment by jdalsem** [comment]
> This PR just feels incomplete, and it provides no overview of what the upgrade path is going to be.

i am trying to make it complete... the upgrade path should become clear (but actually there will be no real upgrade path)...



> It doesn't eliminate all reference to $site_guid, so it feels like it hasn't been thought through

actually it does eliminate all references to $site_guid where it is related to the site_guid of entities... $site_guid is still present when referring to the site_entity... and for the hacky stuff regarding datalist and configtable... the plan is to merge those in another PR... unmanageable (and unneeded) to try to do that as part of this PR



> What's going to happen with config table to that has $site_guid in some cases and not others?

config table always has site_guid... datalist has not. The assumption is that there is only 1 site in config_table. No upgradepath is provided for multisite (and thus for site_guids in entities, acls, config, api users and systemlog)

**Comment by jdalsem** [comment]
I think this is finished... please test and let me know

the only thing left (semi) intact is the usage of site_guid in the elgg_get/save_config functions. This is needed to differentiate between fetching values from datalist or config table. This should be removed as part of the datalist/configtable merger. #2363 We can work on that after this has been merged.

**Comment by jdalsem** [comment]
BE ADVISED: can't return to 2.x after DB is migrated.. so backup before testing

**Comment by mrclay** [comment]
I'll try to get to this soon.

**Pull request merged**